File: javascript/util/publisher.js
Faye.Publisher = {
  countSubscribers: function(eventType) {
    if (!this._subscribers || !this._subscribers[eventType]) return 0;
    return this._subscribers[eventType].length;
  },
  
  addSubscriber: function(eventType, listener, context) {
    this._subscribers = this._subscribers || {};
    var list = this._subscribers[eventType] = this._subscribers[eventType] || [];
    list.push([listener, context]);
  },
  
  removeSubscriber: function(eventType, listener, context) {
    if (!this._subscribers || !this._subscribers[eventType]) return;
    
    var list = this._subscribers[eventType],
        i    = list.length;
    
    while (i--) {
      if (listener && list[i][0] !== listener) continue;
      if (context && list[i][1] !== context) continue;
      list.splice(i,1);
    }
  },
  
  publishEvent: function() {
    var args = Array.prototype.slice.call(arguments),
        eventType = args.shift();
    
    if (!this._subscribers || !this._subscribers[eventType]) return;
    
    Faye.each(this._subscribers[eventType], function(listener) {
      listener[0].apply(listener[1], args.slice());
    });
  }
};

File: lib/faye/publisher.rb
module Faye
  module Publisher
    
    def count_subscribers(event_type)
      return 0 unless @subscribers and @subscribers[event_type]
      @subscribers[event_type].size
    end
    
    def add_subscriber(event_type, listener)
      @subscribers ||= {}
      list = @subscribers[event_type] ||= []
      list << listener
    end
    
    def remove_subscriber(event_type, listener)
      return unless @subscribers and @subscribers[event_type]
      @subscribers[event_type].delete_if(&listener.method(:==))
    end
    
    def publish_event(event_type, *args)
      return unless @subscribers and @subscribers[event_type]
      @subscribers[event_type].each do |listener|
        listener.call(*args)
      end
    end
    
  end
end

